python/pyarrow/_flight.pyx

```diff
@@ -0,0 +1,454 @@
+        return result
+
+    @staticmethod
+    def for_command(command):
+        """Create a FlightDescriptor for an opaque command."""
+        cdef FlightDescriptor result = \
+            FlightDescriptor.__new__(FlightDescriptor)
+        result.descriptor.type = DescriptorTypePath
+        result.descriptor.cmd = tobytes(command)
+        return result
+
+    def descriptor_type(self):
+        # TODO: create a real enum? (wouldn't work in Python < 3.4
+        # without additional dependency)
+        if self.descriptor.type == DescriptorTypeUnknown:
+            return "unknown"
+        elif self.descriptor.type == DescriptorTypePath:
+            return "path"
+        elif self.descriptor.type == DescriptorTypeCmd:
+            return "command"
+        raise RuntimeError("Invalid descriptor type!")
+
+    def command(self):
+        """Get the command for this descriptor."""
+        if self.descriptor_type() != "command":
+            # TODO: raise exception instead?
+            return None
+        return self.descriptor.cmd
+
+    def path(self):
+        """Get the path for this descriptor."""
+        if self.descriptor_type() != "path":
+            return None
+        return self.descriptor.path
+
+    def __repr__(self):
+        return "<FlightDescriptor type: {!r}>".format(self.descriptor_type())
+
+
+class Ticket:
+    """A ticket for requesting a Flight stream."""
+    def __init__(self, ticket):
+        self.ticket = ticket
+
+
+class Location:
+    """A location where a Flight stream is available."""
+    def __init__(self, host, port):
+        self.host = host
+        self.port = port
+
+    def __repr__(self):
+        return '<Location host={} port={}>'.format(self.host, self.port)
+
+
+cdef class FlightEndpoint:
+    """A Flight stream, along with the ticket and locations to access it."""
+    cdef:
+        CFlightEndpoint endpoint
+
+    @staticmethod
+    def create(ticket, locations):
+        cdef:
+            CFlightEndpoint c_endpoint = CFlightEndpoint()
+            CLocation c_location = CLocation()
+
+        if isinstance(ticket, Ticket):
+            c_endpoint.ticket.ticket = ticket.ticket
+        else:
+            c_endpoint.ticket.ticket = ticket
+
+        for location in locations:
+            c_location.host = location.host
+            c_location.port = location.port
+            c_endpoint.locations.push_back(c_location)
+
+        result = FlightEndpoint()
+        result.endpoint = c_endpoint
+        return result
+
+    def ticket(self):
+        return Ticket(self.endpoint.ticket.ticket)
+
+    def locations(self):
+        return [Location(location.host, location.port)
+                for location in self.endpoint.locations]
+
+
+cdef class FlightInfo:
+    """A description of a Flight stream."""
+    cdef:
+        unique_ptr[CFlightInfo] info
+
+    @staticmethod
+    def create(Schema schema, FlightDescriptor descriptor, endpoints,
```

Comment:
Why not make this the `FlightInfo` constructor instead?

In any case, you should add a docstring.